hudi-common/src/main/java/org/apache/hudi/common/util/collection/ExternalSpillableMap.java

```diff
@@ -78,41 +78,49 @@ public class ExternalSpillableMap<T extends Serializable, R extends Serializable
   // Enables compression of values stored in disc
   private final boolean isCompressionEnabled;
   // current space occupied by this map in-memory
-  private Long currentInMemoryMapSize;
+  private long currentInMemoryMapSize;
   // An estimate of the size of each payload written to this map
   private volatile long estimatedPayloadSize = 0;
   // Base File Path
   private final String baseFilePath;
 
-  public ExternalSpillableMap(Long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
+  public ExternalSpillableMap(long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
                               SizeEstimator<R> valueSizeEstimator) throws IOException {
     this(maxInMemorySizeInBytes, baseFilePath, keySizeEstimator, valueSizeEstimator, DiskMapType.BITCASK);
   }
 
-  public ExternalSpillableMap(Long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
+  public ExternalSpillableMap(long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
                               SizeEstimator<R> valueSizeEstimator, DiskMapType diskMapType) throws IOException {
     this(maxInMemorySizeInBytes, baseFilePath, keySizeEstimator, valueSizeEstimator, diskMapType, false);
   }
 
-  public ExternalSpillableMap(Long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
+  public ExternalSpillableMap(long maxInMemorySizeInBytes, String baseFilePath, SizeEstimator<T> keySizeEstimator,
                               SizeEstimator<R> valueSizeEstimator, DiskMapType diskMapType, boolean isCompressionEnabled) throws IOException {
     this.inMemoryMap = new HashMap<>();
     this.baseFilePath = baseFilePath;
-    this.maxInMemorySizeInBytes = (long) Math.floor(maxInMemorySizeInBytes * sizingFactorForInMemoryMap);
+    this.maxInMemorySizeInBytes = (long) Math.floor(maxInMemorySizeInBytes * SIZING_FACTOR_FOR_IN_MEMORY_MAP);
     this.currentInMemoryMapSize = 0L;
     this.keySizeEstimator = keySizeEstimator;
     this.valueSizeEstimator = valueSizeEstimator;
     this.diskMapType = diskMapType;
     this.isCompressionEnabled = isCompressionEnabled;
   }
 
+  private DiskMap<T, R> getDiskBasedMap() {
+    return getDiskBasedMap(false);
+  }
+
+  private DiskMap<T, R> getOrCreateDiskBasedMap() {
+    return getDiskBasedMap(true);
+  }
+
   private DiskMap<T, R> getDiskBasedMap(boolean forceInitialization) {
     if (null == diskBasedMap) {
-      if (!forceInitialization) {
-        return DiskMap.empty();
-      }
       synchronized (this) {
         if (null == diskBasedMap) {
+          if (!forceInitialization) {
+            return DiskMap.empty();
```

Comment:
I can really get the point why `forceInitialization` is needed? Can we just instantiate the `diskBasedMap` when it is null ?

Reply:
You will end up creating the DiskMap even when doing calls like `isEmpty` which introduces more overhead for basic operations when all data is fit in the in-memory map.

Reply:
Can we just optimize the logic in `isEmpty` separately, like to return early if `diskBasedMap` is null.

Reply:
@danny0405 there are about 15 places that suffer from the same issue. I was just using `isEmpty` as a simple example. Check this commit to see just how many places require an update https://github.com/apache/hudi/pull/10344/commits/1c15287300d2ff7385872c1af62017f8e021e9c6

Reply:
Does the empty disk map really makes sense here:

```java
private DiskMap<T, R> getDiskBasedMap() {
    return getDiskBasedMap(false);
  }
```

We see the `put` and `get` both incoke this method, if it returns an empty disk map, the `put` would throw exception. What's the cost there if we just always instantiate an real disk map there, if it is empty, should not cost much resource right?

Reply:
Ok this conversation is going in circles. Can you and @nsivabalan align on the semantics you want? Previously I made all the calls explicit about whether they wanted to force initialization of the map and then I was asked to change that. I think it makes more sense to make it explicit personally since these are all internal methods so there doesn't need to be another layer of abstraction here that can cause confusion.

@danny0405 I think you are missing some of the details here. The outside caller does not need to worry since all the implementation details are hidden from them as they should be. The calls in `put` will only instantiate the map when required. You don't need to instantiate a map to check if a key exists for example. If the map is not instantiated previously, the key will not be present in it.

The cost of instantiating a map on every use is much more expensive. You are doing IO and allocating more objects. If you use RocksDB as your map type, you're going to create column family and all sorts of resources ([source](https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/util/collection/RocksDBDAO.java#L91)). If you use BitCaskDiskMap you are creating file handles and more objects which is lighter weight but still non-zero impact.

I believe this code was intended to have a lazy initialization but was just not implemented correctly.

Reply:
> The calls in put will only instantiate the map when required. You don't need to instantiate a map to check if a key exists for example.

Can you explain the behavior when the first record is put into the disk map, what map can we get from `getDiskBasedMap()` ?

> The cost of instantiating a map on every use is much more expensive. 

Maybe, but personally I don't think it deserves to introduce such complexity.

Reply:
In the `put` operation, you will check if the external map contains the key. If it does, that implies that the external map is already non-empty. Later I use the `getOrCreate` option which will return an existing `DiskMap` or create one as the naming implies. https://github.com/apache/hudi/pull/10344/files#diff-90c0ac84504ff5e04c9de021e9b4f3d14ba092c1e315143764c058c46af21052R236

If the concern is around someone coming into the internals of this class and messing up the logic, then I would suggest my initial approach where you force the caller to acknowledge whether they need to force initialization. It is more verbose but you make the next developer acknowledge that they need to make a choice.

This code also has unit tests which will catch regressions in the logic. Given how prevalent the usage is of this class, I think it makes sense to make it lightweight. For example, the code will setup RocksDB on every instance of `HoodieMergedLogRecordScanner` regardless of the number of records. In my opinion, this small logic update is worth the CPU cycles saved in the creation and closing of these DBs.

Reply:
What `#contains` check are you referring to:

```java
@Override
  public R put(T key, R value) {
    if (this.estimatedPayloadSize == 0) {
      // At first, use the sizeEstimate of a record being inserted into the spillable map.
      // Note, the converter may over-estimate the size of a record in the JVM
      this.estimatedPayloadSize = keySizeEstimator.sizeEstimate(key) + valueSizeEstimator.sizeEstimate(value);
    } else if (this.inMemoryMap.size() % NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE == 0) {
      this.estimatedPayloadSize = (long) (this.estimatedPayloadSize * 0.9 + (keySizeEstimator.sizeEstimate(key) + valueSizeEstimator.sizeEstimate(value)) * 0.1);
      this.currentInMemoryMapSize = this.inMemoryMap.size() * this.estimatedPayloadSize;
    }

    if (this.inMemoryMap.containsKey(key)) {
      this.inMemoryMap.put(key, value);
    } else if (this.currentInMemoryMapSize < this.maxInMemorySizeInBytes) {
      this.currentInMemoryMapSize += this.estimatedPayloadSize;
      // Remove the old version of the record from disk first to avoid data duplication.
      if (inDiskContainsKey(key)) {
        getDiskBasedMap().remove(key);
      }
      this.inMemoryMap.put(key, value);
    } else {
      getDiskBasedMap().put(key, value);
    }
    return value;
  }
```

Reply:
`inDiskContainsKey(key))` which references the DiskMap

Reply:
This line is to fix the duplicates between in-momory and disk-based map. But `inDiskContainsKey(key)` will just return `EmptyMap` if we only got in-memory records, when `this.currentInMemoryMapSize >= this.maxInMemorySizeInBytes`, the code goes to `getDiskBasedMap().put(key, value);`, and still we put records into an `EmptyMap`.

Reply:
Where does the code call `getDiskBasedMap().put(key, value)` in this branch? All unit tests are passing so we cannot be calling put on empty map as that would throw an exception.

Reply:
https://github.com/apache/hudi/pull/10344/files#diff-90c0ac84504ff5e04c9de021e9b4f3d14ba092c1e315143764c058c46af21052R236

Reply:
Here:

```java
```

Reply:
Looks like the link did not paste correctly

Reply:
@danny0405 i see you are referring to the code on master. What is the relevance? I have linked the same section above but it is from this PR. 

I can check the tests tomorrow but I really want to understand what I can do to resolve the issues around the approach. I have offered my opinion on how to make it clear but I have not gotten feedback on that.

Reply:
Yeah, my bad, if `getOrCreateDiskBasedMap()` is used, then the `#put` should work correctly.

Reply:
It has been in use since before this thread of comments started. I'm not sure I understand what is going on at this point but let me know how you think the code can be made more clear.

Reply:
My biggest concern is for write method (like `#put`) and read method (like `#contains`) we need to take care the disk type to ensure the correctness.

Another choice is we make the instantiation of `RocksDbDiskMap` and `BitCaskDiskMap` as lazy and light-weight, so that the wrapper `ExternalSpillableMap` can simplify a lot.

Reply:
> My biggest concern is for write method (like `#put`) and read method (like `#contains`) we need to take care the disk map type to ensure the correctness.
> 
Yes, this is what unit tests are for. Once we agree on a path, I will make sure they are in place and handle the various code paths in the `put` and `contains` logic.

> Another choice is we make the instantiation of `RocksDbDiskMap` and `BitCaskDiskMap` as lazy and light-weight, so that the wrapper `ExternalSpillableMap` can simplify a lot.
Wouldn't this just duplicate the logic?  If someone adds a third implementation, then would it be triple work? How do you propose doing this in a way that has less complexity? This complexity bit seems to be the major blocker so I am looking for some clear ways on how to actually reduce it. Links to where code should be added would be helpful.

Reply:
I will sync up w/ danny directly and get some consensus.

Reply:
> Wouldn't this just duplicate the logic? If someone adds a third implementation, then would it be triple work? How do you propose doing this in a way that has less complexity? 

I believe we can abstract it out as common in the base class.

Reply:
@danny0405 let's breakdown this suggestion. You are asking to:
1. Introduce a new abstract init method in `DiskMap` 
2. Move the class initialization of DiskMap to its own init method that 
3. Make sure the initialization methods are only called once there is a `put` called
4. In all other methods in DiskMap, account for whether or not the init method has been called.
5. Add unit tests or some test suite for DiskMap and its implementations to ensure all future iterations properly implement the lazy initialization.

The changes I have in the `ExternalSpillableMap` map are around 30 lines. There is an existing set of tests that can be augmented to make sure that the logic is working properly. 

I want to bring up what you have claimed is your biggest concern with the PR:
```
My biggest concern is for write method (like #put) and read method (like #contains) we need to take care the disk map type to ensure the correctness.
```
If this is your biggest concern, why isn't a unit test sufficient? @danny0405 I have already suggested this much and you have not addressed whether this is sufficient.

Reply:
Inside of the [existing tests](https://github.com/apache/hudi/blob/master/hudi-common/src/test/java/org/apache/hudi/common/util/collection/TestExternalSpillableMap.java) there are assertions on `containsKey` if that is what you mean by `contains`. These tests also add elements with `put` so we have testing around this already.

Reply:
> If this is your biggest concern, why isn't a unit test sufficient

It's not about correctness, it's about the maintainability of the code.

Reply:
Ok I've made suggestions on this matter as well that I feel have gone ignored. 

Also this is a map interface and the methods are already implemented. The amount of new development should be minimal at this point since it would not make sense to add new methods. Don't you agree?

Reply:
I'm wondering if we can make the two map implementations initiaze lazily by themselves so that in this ExternalSpillableMap there is no need to pay special attention to make the read/write mehods as corrrect, and there is no need to introduce the dummy empty map.

Reply:
At some point you will need to pay special attention to whether the read/write methods are correct. Right now we are just debating about where that is, is that correct?

In my opinion, the ExternalSpillableMap is the logical place for handling the logic of initializing the disk map if it is needed. We can avoid the dummy empty map by also embedding null checks into all of the methods.

Reply:
> We can avoid the dummy empty map by also embedding null checks into all of the methods

Somehow makes sense, I just thought it might be more straight-forward to do that in specific map impls.